### PR TITLE
 Prototype repo plugin for AnVIL workspaces in Terra (#3804)

### PR DESCRIPTION
**Pull request by noah-aviel-dove**
https://github.com/DataBiosphere/azul/issues/3804

### Author

- [x] PR title references issue
- [x] PR title matches issue title (preceded by `Fix: ` for bugs)   <sub>or there is a good reason why they're different</sub>
- [x] Title of main commit references issue
- [x] PR is connected to Zenhub issue and description links to issue

### Author (reindex)

- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>

### Author (freebies & chains)

- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in the chain        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>

### Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>
- [x] Added announcement to PR description                  <sub>or this PR does not require announcement</sub>
- [x] Added checklist items for additional operator tasks   <sub>or this PR does not require additional tasks</sub>

### Author (requirements, before every review)

- [x] Ran `make requirements_update`                        <sub>or this PR does not touch requirements*.txt, common.mk, Makefile and Dockerfile</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR does not touch requirements*.txt</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR does not touch requirements*.txt</sub>

### Author (before every review)

- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

### Primary reviewer (after approval)

- [ ] Commented in issue about demo expectations            <sub>or labelled issue as `no demo`</sub>
- [ ] Decided if PR can be labeled `no sandbox`
- [ ] PR title is appropriate as title of merge commit
- [ ] Moved ticket to Approved column
- [ ] Assigned PR to an operator

### Operator (before pushing merge the commit)

- [ ] Checked `reindex` label and `r` commit title tag
- [ ] Checked that demo expectations are clear              <sub>or issue is labeled as `no demo`</sub>
- [ ] Rebased and squashed branch
- [ ] Sanity-checked history
- [ ] Pushed PR branch to Github
- [ ] Branch pushed to Gitlab and added `sandbox` label     <sub>or PR is labeled `no sandbox`</sub>
- [ ] Build passed in sandbox                               <sub>or PR is labeled `no sandbox`</sub>
- [ ] Started reindex in `sandbox`                          <sub>or this PR does not require reindexing `sandbox`</sub>
- [ ] Checked for failures in `sandbox`                     <sub>or this PR does not require reindexing `sandbox`</sub>
- [ ] Added PR reference to merge commit title
- [ ] Collected commit title tags in merge commit title
- [ ] Moved linked issue to Merged column
- [ ] Pushed merge commit to Github

### Operator (after pushing the merge commit)

- [ ] Made announcement requested by author                 <sub>or PR description does not contain an announcement</sub>
- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened the PR chain                                <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate       <sub>or this PR is authored by lead</sub>
- [ ] Pushed merge commit to Gitlab                         <sub>or merge commit can be pushed later, with another PR</sub>
- [ ] Deleted PR branch from Github and Gitlab
- [ ] Build passes on Gitlab
- [ ] Moved issues to `prod` or `Merged prod`               <sub>or this PR does not represent a promotion</sub>

### Operator (reindex) 

- [ ] Started reindex in target deployment                  <sub>or this PR does not require reindexing</sub>
- [ ] Checked for and triaged indexing failures             <sub>or this PR does not require reindexing</sub>
- [ ] Emptied fail queues in target deployment              <sub>or this PR does not require reindexing</sub>
- [ ] Filed backport PR                                     <sub>or this PR does not represent a hotfix</sub>

### Operator

- [ ] Unassigned PR

### CODE REVIEW

**Comment by noah-aviel-dove** [comment]
Reviewers: note the `drop!` commit. The implementation of the new repository plugin is incomplete so some functionality has to be skipped for the IT to pass. However since no new catalogs have been added that use the new plugin, we don't actually need to tweak the IT for deployments besides my own. IT takes over 1.5 hours to pass on the `drop!` commit due to the slow response times from the Terra workspace API.

File: src/azul/plugins/repository/terra/__init__.py [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+    pass
+
+
+TerraBundleFQID = SourcedBundleFQID[TerraSourceRef]
+
+
+class TerraBundle(Bundle):
+
+    def drs_path(self, manifest_entry: JSON) -> Optional[str]:
+        assert manifest_entry.get('drs_Path') is None, manifest_entry
```
Comment (danielsotirhos) [inline src/azul/plugins/repository/terra/__init__.py:60]:
```suggestion
        assert manifest_entry.get('drs_path') is None, manifest_entry
```

**Review by danielsotirhos — CHANGES_REQUESTED** [review]
The rest looks great. Please rebase to resolve conflicts.

**Review by danielsotirhos — APPROVED** [review]
Approved.

**Comment by melainalegaspi** [comment]
@hannes-ucsc : "I haven't reviewed this yet but I think we should prepare for a quick demo that involves the Data Browser. Could you please set up a locally running DB instance that's backed by your AnVIL-specific personal deployment of Azul."

**Pull request closed without merging**